### PR TITLE
Tracks: Add Ability to Filter Event Props

### PR DESCRIPTION
**Pull request by timmyc**
This branch seeks to add two new filters to the tracks utilities in WooCommerce. For full details behind this proposed change, please see: p7bje6-1Uy-p2

But TL;DR; this change allows us, or anyone to modify the event properties of usage tracks within WooCommerce and WooCommerce Admin functionality.

To ease in testing out this PR, I've create [an example plugin](https://gist.github.com/timmyc/edf820bded9a805a0746751a421bfb8a) that implements both track event properties filters, and has some logging for you to see things in action.

__To Test PHP Filter__
- Download and activate the example plugin: https://gist.github.com/timmyc/edf820bded9a805a0746751a421bfb8a
- tail your PHP error log file, and then navigate to the order listing page
- verify you see the following in your log file:

```
[13-Mar-2020 23:46:44 UTC] Event Name: wcadmin_orders_view
[13-Mar-2020 23:46:44 UTC] Array
(
    [status] => all
    [host] => isorex
    [amaze] => stuff
)
```

- Open an order, and edit it  ( i.e. change the status )
- Verify you see something like the following in your log:

```
[13-Mar-2020 23:48:44 UTC] Event Name: wcadmin_order_edit_date_created
[13-Mar-2020 23:48:44 UTC] Array
(
    [order_id] => 2594
    [status] => completed
    [host] => isorex
)

[13-Mar-2020 23:48:45 UTC] Event Name: wcadmin_orders_edit_status_change
[13-Mar-2020 23:48:45 UTC] Array
(
    [order_id] => 2594
    [next_status] => cancelled
    [previous_status] => completed
    [date_created] => 2020-03-06
    [payment_method] => cod
    [host] => isorex
)
```

__To Test JS Filter__
- With the example plugin installed, open up dev tools in your browser to the console
- Visit the status page `wp-admin/admin.php?page=wc-status`
- Click on the "Get System Report" button:
image
- Verify you see the following in your JS console:
image

### CODE REVIEW

**Comment by haszari** [comment]
> This branch seeks to add two new filters to the tracks utilities in WooCommerce.
> TL;DR; this change allows us, or anyone to modify the event properties of usage tracks within WooCommerce and WooCommerce Admin functionality.

Great to have a hook to allow adding additional data into tracks events :)

I'm wondering if it's technically possible add props to all tracks events from one hook. For example, perhaps the PHP hook which could enqueue the extra props for the front end to append when sending events? 

Having two different hooks for the two tracks APIs seems like a technical detail, and the normal case would be to add the same props to all tracks events - so it would be extra nice if there was one hook to implement to achieve this.

I can see that having filters for JS and PHP allows the props to be dynamic, based on the event - though is that flexibility needed?

**Review by psealock — COMMENTED** [review]
This is all working and testing well @timmyc ! 

Just a few comments, but I did notice something odd. The error is definitely in https://github.com/Automattic/analytics/blob/master/lib/w/tracks.js, but this filter allows plugins to manipulate `_ui` and `_ut` properties. You can add those parameters in the plugin and view the t.gif request query parameters. Its probably not a thing, but it would be safest to either fix the implementation 🙀 or just filter out those properties if they exist from `eventProperties`.

File: includes/tracks/class-wc-site-tracking.php [hunk 1]
```diff
@@ -68,6 +68,10 @@ public static function add_tracking_function() {
 			window.wcTracks.recordEvent = function( name, properties ) {
 				var eventName = '<?php echo esc_attr( WC_Tracks::PREFIX ); ?>' + name;
 				var eventProperties = properties || {};
+				if ( window.wp && window.wp.hooks && window.wp.hooks.applyFilters ) {
+					eventProperties = window.wp.hooks.applyFilters( 'woocommerceTracksEventProperties', eventProperties, eventName );
+					console.log( 'new', eventProperties );
+				}
 				eventProperties.url = '<?php echo esc_html( home_url() ); ?>'
 				eventProperties.products_count = '<?php echo intval( WC_Tracks::get_products_count() ); ?>';
```
[the remark this answers is not in the document]
Comment (timmyc) [reply includes/tracks/class-wc-site-tracking.php]: oh great idea.

**Comment by timmyc** [comment]
> but this filter allows plugins to manipulate _ui and _ut properties

Thanks @psealock definitely was wondering how these filters could potentially cause harm. Any other props do you think we need to add to a blacklist to not allow manipulation?

**Comment by psealock** [comment]
>Any other props do you think we need to add to a blacklist to not allow manipulation?

The rest are good. Attempts to write to them get overwritten by the Tracks js, but `_ut` and `_ui` are different because the logic looks for those properties and uses them if they are present.

**Comment by psealock** [comment]
>Having two different hooks for the two tracks APIs seems like a technical detail, and the normal case would be to add the same props to all tracks events - so it would be extra nice if there was one hook to implement to achieve this.

Interesting idea, I think it would be technically possible. It would be a PHP based filter and there'd be no dynamic filtering.

**Comment by timmyc** [comment]
yeah interesting idea @haszari - we could consolidate the two, and it would just apply global props to all tracks. Like your event props would get exported to JSON, and then the JS based tracks would just use that.

As @psealock said this would limit the ability to hook in further custom logic based upon certain event names, but it would definitely satisfy the need for global props - but it would definitely fulfill the immediate use case.

From an implementation standpoint, the current approach seems pretty easy to work with. Yes the two contexts aren't completely ideal, but it does offer a bit more flexibility to only manipulate props on certain events if desired.

Thanks for the feedback on the PR Paul - I applied the updates here. I also changed the PHP filter name to comply with underscores instead of dashes.

**Review by psealock — APPROVED** [review]
This is looking great @timmyc. Pre-approving, but just to point out that our js based filter names are underscore also:

https://github.com/woocommerce/woocommerce-admin/blob/e92964532bdadfbea8d6d7bdfb5598c2b1a59a57/client/analytics/report/orders/config.js#L17

**Comment by timmyc** [comment]
With the code freeze coming up, hoping to get a final set of eyes from Proton on this prior to merge /cc @vedanshujain

File: includes/tracks/class-wc-site-tracking.php [hunk 2]
```diff
@@ -70,6 +70,11 @@ public static function add_tracking_function() {
 				var eventProperties = properties || {};
 				eventProperties.url = '<?php echo esc_html( home_url() ); ?>'
 				eventProperties.products_count = '<?php echo intval( WC_Tracks::get_products_count() ); ?>';
+				if ( window.wp && window.wp.hooks && window.wp.hooks.applyFilters ) {
+					eventProperties = window.wp.hooks.applyFilters( 'woocommerce_tracks_client_event_properties', eventProperties, eventName );
+					delete( eventProperties._ui );
```
Comment (claudiosanches) [inline includes/tracks/class-wc-site-tracking.php]: It is an operator, so no need for `()`.

**Review by claudiosanches — APPROVED** [review]
LGTM!

Just a minor coding standards issue.

**Comment by vedanshujain** [comment]
(Note: Merging despite coding standard issue as @claudiosanches mentioned because it does not seem to fail standard tests. Perhaps consider fixing standard voilation in a new PR cc @timmyc )
